[18.0][MIG] connector_importer_source_sftp: Migration to 18.0

yankinmax opened the pull request.

simahawk committed:
Add connector_importer_source_sftp

simahawk committed:
connector_importer_source_sftp: migrate to v14

damdam-s committed:
c_importer_source_sftp: fix move file

A file can be moved only if we have a destination path.

dhx committed:
[MIG] connector_importer_source_sftp: Migration to 15.0

dhx committed:
[IMP] connector_importer_source_sftp: Handle no or empty file found case

Instead of throwing an exception with an unclear message in the _get_line function

simahawk committed:
connector_importer_source_sftp: fix pin cryptography<37

yankinmax committed:
[MIG] connector_importer_source_sftp: Migration to 18.0

yankinmax commented:
Hello @simahawk @sebalix @SilvioC2C can you pls help with any suggestions?
I'm not familiar with the module usage.
My debug session shows that `filelist = self.list(relative_path)` is empty, so no files are really found.
https://github.com/OCA/storage/blob/5bbfe12c01e68e562922838aa5fd0eb28dc235a0/storage_backend/components/base_adapter.py#L32

yankinmax commented:
> Hello @simahawk @sebalix @SilvioC2C can you pls help with any suggestions? I'm not familiar with the module usage. My debug session shows that `filelist = self.list(relative_path)` is empty, so no files are really found. https://github.com/OCA/storage/blob/5bbfe12c01e68e562922838aa5fd0eb28dc235a0/storage_backend/components/base_adapter.py#L32

So, the issue is definitely with the `_selection_source_ref_id` override to include `import.source.csv.sftp` source.
It's not called somehow :cry: 

https://github.com/OCA/connector-interfaces/pull/172/files#r2476825551

yankinmax commented on connector_importer_source_sftp/models/source_mixin.py:
```diff
@@ -0,0 +1,19 @@
+    _inherit = "import.source.consumer.mixin"
+
+    @api.model
+    def _selection_source_ref_id(self):
+        selection = super()._selection_source_ref_id()
+        new_source = ("import.source.csv.sftp", "CSV SFTP")
+        if new_source not in selection:
+            selection.append(new_source)
+        return selection
```
this is not working as expected

yankinmax commented on connector_importer_source_sftp/models/__init__.py:
```diff
@@ -1,2 +1,2 @@
-from . import source_csv_sftp
 from . import source_mixin
+from . import source_csv_sftp
```
@ivantodorovich this idea didn't help

ivantodorovich commented on connector_importer_source_sftp/models/source_csv_sftp.py:
```diff
@@ -0,0 +1,133 @@
+
+    _name = "import.source.csv.sftp"
+    _inherit = [
+        "import.source.csv",
```
try also inheriting from the mixin again here


```suggestion
        "import.source.csv",
        "import.source.consumer.mixin",
```

yankinmax replied:
@ivantodorovich something really unexpected happening here. Seems still not working as expected. I wonder if we could adapt to use `fields.Selection` here?

ivantodorovich replied:
Yes, simplify if needed.

yankinmax replied:
Making `source_model` selection attribute as usual list of tuples and adding to the inherited class `selection_add` also doesn't work.

yankinmax replied:
@simahawk @sebalix do you have any idea of the issue that `_selection_source_ref_id` override doesn't actually add `new_source`, simply not called.

`connector_importer_source_sftp/models/source_mixin.py`
```
@api.model
def _selection_source_ref_id(self):
    selection = super()._selection_source_ref_id()
    new_source = ("import.source.csv.sftp", "CSV SFTP")
    if new_source not in selection:
        selection.append(new_source)
    return selection
```

ivantodorovich committed:
[FIX] connector_importer: source consumer mixin

Traceback:
```
  File ".../connector_importer_source_sftp/components/event_listeners.py", line 32, in _skip_move_file
    source = importer.recordset.get_source()
  File ".../connector_importer/models/sources/source_consumer_mixin.py", line 71, in get_source
    return self.source_ref_id
  File "/opt/odoo/odoo/fields.py", line 1303, in __get__
    self.compute_value(recs)
  File "/opt/odoo/odoo/fields.py", line 1485, in compute_value
    records._compute_field_value(self)
  File "/opt/odoo/odoo/models.py", line 5297, in _compute_field_value
    fields.determine(field.compute, self)
  File "/opt/odoo/odoo/fields.py", line 110, in determine
    return needle(*args)
  File ".../connector_importer/models/sources/source_consumer_mixin.py", line 43, in _compute_source_ref_id
    item.source_ref_id = f"{item.source_model},{item.source_id}"
  File "/opt/odoo/odoo/fields.py", line 1398, in __set__
    self.write(protected_records, value)
  File "/opt/odoo/odoo/fields.py", line 1205, in write
    cache_value = self.convert_to_cache(value, records)
  File "/opt/odoo/odoo/fields.py", line 3041, in convert_to_cache
    raise ValueError("Wrong value for %s: %r" % (self, value))
ValueError: Wrong value for import.recordset.source_ref_id: 'import.source.csv.sftp,1'
```

For some reason I don't know, by the time the event listener skip_if condition is
evaluated, the `import.recordset` model does not have applied the override to
`_selection_source_ref_id` in the mixin, and so the Reference field fails with
this ValueError.

The workaround here makes `get_source` not actually use the Reference field, but
directly `source_model` and `source_id`, which of course do not attempt to
validate the value.

ivantodorovich committed:
[FIX] connector_importer_source_sftp: server env issue in demo data & tests

For some reason, probably due to server_environment, if tests are ran during the
module installation, like in the oca-ci, the storage backend_type is
"filesystem", when it should be "sftp" according to the demo data.

Did not investigate any further.

yankinmax commented:
Thanks for help @ivantodorovich :pray:

ivantodorovich committed:
[IMP] connector_importer_source_sftp: properly raise exceptions if file not found

ivantodorovich commented:
@simahawk this one's ready too

simahawk commented:
/ocabot merge nobump
